lib/evaluate/constant.cc

```diff
@@ -30,22 +30,59 @@ std::size_t TotalElementCount(const ConstantSubscripts &shape) {
 }
 
 bool IncrementSubscripts(ConstantSubscripts &indices,
-    const ConstantSubscripts &shape, const ConstantSubscripts &lbound) {
+    const ConstantSubscripts &shape, const ConstantSubscripts &lbound,
+    const std::vector<int> *dimOrder) {
   int rank{GetRank(shape)};
   CHECK(GetRank(indices) == rank);
+  CHECK(GetRank(lbound) == rank);
+  CHECK(!dimOrder || static_cast<int>(dimOrder->size()) == rank);
   for (int j{0}; j < rank; ++j) {
     auto lb{lbound[j]};
-    CHECK(indices[j] >= lb);
-    if (++indices[j] < lb + shape[j]) {
+    ConstantSubscript k{dimOrder ? (*dimOrder)[j] : j};
+    CHECK(indices[k] >= lb);
+    if (++indices[k] < lb + shape[k]) {
       return true;
     } else {
-      CHECK(indices[j] == lb + shape[j]);
-      indices[j] = lb;
+      CHECK(indices[k] == shape[k] + lb);
+      indices[k] = lb;
     }
   }
   return false;  // all done
 }
 
+std::optional<std::vector<int>> IsValidDimensionOrder(
```

Comment:
I feel the name gives a wrong impression about the return value of the function.

Especially since there are other functions of the form `IsValidShape` which returns `bool`.

Reply:
I noticed that you have a comment next to this function's signature. I suppose that's okay

Reply:
`ValidateDimensionOrder` might be a better name.

Also, it's unfortunate that this function uses `std::optional<std::vector<int>>` in its interface while others in this file use `const std::vector<int> *`. Should the others, like `CopyFrom`, use `const std::optional<std::vector<int>> &` for `dimOrder`?

Reply:
Thanks, the name was indeed not that good, I took `ValidateDimensionOrder`.
For the interface aspect, it was more straight forward on the call site in that case, but I do not want to have  an optional `optional` argument in `IncrementSubscripts`. 

I actually completely removed the optional aspect on `ValidateDimensionOrder`, it did not fit with the name and it did not save that much (I also reorganized the if imbrication at the call site to use an if/else-if structure that is more straightforward).